app/gui/view/graph-editor/src/selection/bounding_box.rs

```diff
@@ -83,31 +144,96 @@ impl BoundingBox {
-        let bb2 = BoundingBox::from_corners(Vector2::new(2.0, 2.0), Vector2::new(200.0, 200.0));
-        assert!(bb1.intersects(&bb2));
-        assert!(bb2.intersects(&bb1));
-
-        let bb1 = BoundingBox::from_corners(Vector2::new(-50.0, -50.0), Vector2::new(25.0, 25.0));
-        let bb2 = BoundingBox::from_corners(Vector2::new(5.00, 50.0), Vector2::new(100.0, 100.0));
-        assert!(!bb1.intersects(&bb2));
-        assert!(!bb2.intersects(&bb1));
+        assert_intersect! { ((0.5, 0.5), (1.0, 1.0))  *?  ((0.0, 0.0), (  2.0,   2.0))  ==  true };
+        assert_intersect! { ((3.0, 3.0), (4.0, 4.0))  *?  ((0.0, 0.0), (  2.0,   2.0))  ==  false };
+        assert_intersect! { ((0.0, 0.0), (4.0, 4.0))  *?  ((0.0, 0.0), ( -2.0,  -2.0))  ==  true };
+        assert_intersect! { ((0.0, 0.0), (4.0, 4.0))  *?  ((2.0, 2.0), (200.0, 200.0))  ==  true };
+        assert_intersect! {
+            ((-50.0, -50.0), (25.0, 25.0))  *?  ((5.00, 50.0), (100.0, 100.0))  ==  false
+        };
+    }
+
+    macro_rules! assert_concat {
+        ( $( $bbox1:tt + $bbox2:tt == $bbox3:tt ; )+ ) => {
+            $( assert_concat!{$bbox1 + $bbox2 == $bbox3}; )+
+        };
+
+        ( $bbox1:tt + $bbox2:tt == $bbox3:tt ) => {
+            let bbox1: BoundingBox = $bbox1.into();
+            let bbox2: BoundingBox = $bbox2.into();
+            let bbox3: BoundingBox = $bbox3.into();
+            let result = bbox1.concat(bbox2);
+            let assert_msg = iformat!(
+                "Concat result was expected to be: " bbox3;? " but got: " result;? " instead.");
+            assert_eq!(result.left, bbox3.left, "{}", assert_msg);
+            assert_eq!(result.right, bbox3.right, "{}", assert_msg);
+            assert_eq!(result.top, bbox3.top, "{}", assert_msg);
+            assert_eq!(result.bottom, bbox3.bottom, "{}", assert_msg);
+        };
+    }
+
+    #[test]
+    fn test_concat() {
+        assert_concat! {
+            ((0.0, 0.0), (2.0, 3.0)) + (( 2.0,  3.0), (4.0, 5.0)) == ((0.0,   0.0), (4.0, 5.0));
+            ((0.0, 0.0), (1.0, 1.0)) + ((-1.0, -1.0), (0.5, 0.5)) == ((-1.0, -1.0), (1.0, 1.0));
+            ((0.0, 0.0), (1.0, 1.0)) + (( 0.3,  0.3), (0.6, 0.6)) == ((0.0,   0.0), (1.0, 1.0));
+        };
+    }
+
+    const SQUARED_DISTANCE_COMPARISON_PRECISION: f32 = 0.001;
+
+    macro_rules! assert_squared_distance_to_point {
+        ( $( $bbox:tt <-sq-> $point:tt =~ $expected_sq_distance:expr ; )+ ) => {
+            $(
+                assert_squared_distance_to_point!{ $bbox <-sq-> $point =~ $expected_sq_distance };
+            )+
+        };
+
+        ($bbox:tt <-sq-> $point:tt =~ $expected_sq_distance:expr) => {
+            let bbox: BoundingBox = $bbox.into();
+            let point = Vector2($point.0, $point.1);
+            let result = bbox.squared_distance_to_point(point);
+            let result_deviation = (result - $expected_sq_distance).abs();
```

Comment:
Could this not be done easier with `assert_approx_eq`?

Reply:
Unfortunately, `assert_approx_eq!` does not allow providing an extra message. An extra message is needed here because the line number of the particular test case is lost due to all the test cases being wrapped in a macro call in: https://github.com/enso-org/enso/blob/e9d031cb057c3a3b4a737c4645e788768a3475aa/app/gui/view/graph-editor/src/selection/bounding_box.rs#L228-L238. All the test cases are wrapped in a macro call to keep them from crossing the line length limit - otherwise, a line like below is over 100 characters long:
```rust
        assert_squared_distance_to_point! { ((-1.0, -1.0), (0.0, 0.0)) <-sq-> ( 3.0, 4.0) =~ 5.0.pow(2.0) };
```

Would you have a suggestion on how to improve this given the constraints of keeping a line length below 100 characters, avoiding multi-line expressions, and [avoiding repetitiveness in tests](https://github.com/enso-org/enso/pull/3301#discussion_r819338814)? Should I implement an `assert_approx_eq` macro in our prelude, copying the one from the third-party package, but adding a variant allowing for passing a custom message?

Reply:
Do they have to be done via macro? Could this not just as well be a helper function taking arguments? Then this would all be taken care of by the auto formatter.
Just a `fn assert_squared_distance_to_point(bbox:BoundingBox,target:Vector2,target_distance:f64)` or, if that makes it more readable: `fn assert_squared_distance_to_point(bbox:[[f64;2];2],target:[f64;2],target_distance:f64)`.

Reply:
Based on discussion in https://discord.com/channels/401396655599124480/949323065320415292/949323743430316062 and our further discussion on Discord with @MichaelMauderer, I currently plan to keep the macros approach here, and @MichaelMauderer asked me to disregard this thread of his comments above.